### PR TITLE
Sponsorship invariant

### PR DESCRIPTION
**Pull request by jonjove**
# Description
Update the invariants subsystem for sponsorships.

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/main/ApplicationImpl.cpp [hunk 1]
```diff
@@ -915,4 +917,4 @@ ApplicationImpl::getLedgerTxnRoot()
     return mConfig.MODE_USES_IN_MEMORY_LEDGER ? *mNeverCommittingLedgerTxn
                                               : *mLedgerTxnRoot;
 }
-}
\ No newline at end of file
+}
```
Comment (sisuresh) [inline src/main/ApplicationImpl.cpp:920]: Nit: Looks like you added a newline?

**Comment by sisuresh** [comment]
@jonjove Should we include some test cases for `SponsorshipCountIsValid`? I noticed that we didn't have any test cases for `LedgerEntryIsValid` and it got me thinking. I think we should have a test case to at least verify that each invariant is being run.

**Comment by jonjove** [comment]
Yes, the reason I've been delaying on merging this is that I haven't gotten around to writing tests. But I think it's a good principle to test that:
a) every invariant is at least getting run (as you point out)
b) it handles some common cases and non-cases correctly

Hopefully I'll get the tests written next week.

**Review by MonsieurNicolas — CHANGES_REQUESTED** [review]

File: src/invariant/LedgerEntryIsValid.cpp [hunk 2]
```diff
@@ -149,6 +154,19 @@ LedgerEntryIsValid::checkIsValid(AccountEntry const& ae, uint32 version) const
             return "Account signers have invalid weights";
         }
     }
+
+    if (ae.ext.v() == 1 && ae.ext.v1().ext.v() == 2)
```
[the remark this answers is not in the document]
Comment (jonjove) [reply src/invariant/LedgerEntryIsValid.cpp]: Fixed.

File: src/ledger/test/LedgerTestUtils.cpp [hunk 3]
```diff
@@ -225,6 +225,11 @@ makeValid(ClaimableBalanceEntry& c)
     c.amount = std::abs(c.amount);
     clampLow<int64>(1, c.amount);
 
+    if (c.claimants.empty())
+    {
+        c.claimants.resize(1);
```
[the remark this answers is not in the document]
Comment (jonjove) [reply src/ledger/test/LedgerTestUtils.cpp:233]: Added a comment.

File: src/invariant/SponsorshipCountIsValid.cpp [hunk 4]
```diff
@@ -0,0 +1,250 @@
+// Copyright 2020 Stellar Development Foundation and contributors. Licensed
```
[the remark this answers is not in the document]
Comment (jonjove) [reply src/invariant/SponsorshipCountIsValid.cpp:1]: Done.

File: src/invariant/test/SponsorshipCountIsValidTests.cpp [hunk 5]
```diff
@@ -0,0 +1,511 @@
+// Copyright 2020 Stellar Development Foundation and contributors. Licensed
```
[the remark this answers is not in the document]
Comment (jonjove) [reply src/invariant/test/SponsorshipCountIsValidTests.cpp:1]: Done.

**Comment by MonsieurNicolas** [comment]
r+ 142fc5914ede400c4517a739aef761a89e12fec7

**Pull request merged**